### PR TITLE
cmd, dot: migrate node logic to dot

### LINKED ISSUE #252: List types for CLI params
Having had issues with ambiguous cli docs in the past, it would be awesome to have consistent input type specifications (number, string, comma-separated list). A simple mapping could append it to cli.Flag.Usage (`cmd/utils/flags.go`)

### PR DESCRIPTION
**Pull request by ryanchristo**
## Changes

- migrate configuration logic from command package to dot package
- migrate keystore logic from command package to keystore package
- create test utils package and implement in cmd and dot root level

## Tests:

```
go clean --testcache ./cmd/...
go test ./cmd/...
```

```
go clean --testcache ./dot
go test ./dot
```

```
go clean --testcache ./lib/keystore
go test ./lib/keystore
```

```
go clean --testcache ./lib/utils
go test ./lib/utils
```

```
go clean --testcache ./...
go test ./...
```

build and run gossamer command:

```
make gossamer
./bin/gossamer init --key alice --datadir tmp/alice --verbosity trace
./bin/gossamer --key alice --datadir tmp/alice --verbosity trace
```

## Issues:

closes #252

### CODE REVIEW

File: cmd/gossamer/account_test.go [hunk 1]
```diff
@@ -17,190 +17,67 @@
...
-	}
 
-	defer os.RemoveAll(testKeystoreDir)
-
-	command := accountCommand
-	err = command.Run(ctx)
-	if err != nil {
-		t.Fatal(err)
-	}
+	// TODO: check contents of data directory - improve cmd account tests
```
Comment (ryanchristo) [inline cmd/gossamer/account_test.go]: #685

File: cmd/gossamer/account_test.go [hunk 2]
```diff
@@ -17,190 +17,67 @@
...
-	}
-
-	if len(keys) != 1 {
-		t.Fatal("Fail: expected 1 key in keystore")
-	}
-
-	if strings.Compare(keys[0], filepath.Base(keyfile)) != 0 {
-		t.Fatalf("Fail: got %s expected %s", keys[0], keyfile)
-	}
+	// TODO: check contents of data directory - improve cmd account tests
```
Comment (ryanchristo) [inline cmd/gossamer/account_test.go]: #685

File: cmd/gossamer/account_test.go [hunk 3]
```diff
@@ -210,50 +87,43 @@ func TestImportCommand(t *testing.T) {
...
 		t.Fatal(err)
 	}
-}
 
-func TestImportKey_ShouldFail(t *testing.T) {
-	_, err := importKey("./notakey.key", testKeystoreDir)
-	if err == nil {
-		t.Fatal("did not err")
-	}
+	// TODO: check contents of data directory - improve cmd account tests
```
Comment (ryanchristo) [inline cmd/gossamer/account_test.go]: #685

File: cmd/gossamer/account_test.go [hunk 4]
```diff
@@ -210,50 +87,43 @@ func TestImportCommand(t *testing.T) {
...
-	defer os.RemoveAll(testKeystoreDir)
-
-	if len(keys) != 1 {
-		t.Fatal("fail")
-	}
-
-	if strings.Compare(keys[0], filepath.Base(keyfile)) != 0 {
-		t.Fatalf("Fail: got %s expected %s", keys[0], keyfile)
-	}
+	// TODO: check contents of data directory - improve cmd account tests
```
Comment (ryanchristo) [inline cmd/gossamer/account_test.go]: #685

File: cmd/gossamer/account_test.go [hunk 5]
```diff
@@ -263,44 +133,6 @@ func TestListCommand(t *testing.T) {
...
-	}
-
-	sort.Slice(expected, func(i, j int) bool { return strings.Compare(expected[i], expected[j]) < 0 })
-
-	for i, key := range keys {
-		if strings.Compare(key, filepath.Base(expected[i])) != 0 {
-			t.Fatalf("Fail: got %s expected %s", key, filepath.Base(expected[i]))
-		}
-	}
+	// TODO: check contents of data directory - improve cmd account tests
```
Comment (ryanchristo) [inline cmd/gossamer/account_test.go]: #685

File: cmd/gossamer/config_test.go [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+	"testing"
+
+	"github.com/ChainSafe/gossamer/dot"
+	"github.com/ChainSafe/gossamer/lib/utils"
+
+	"github.com/stretchr/testify/require"
+	"github.com/urfave/cli"
+)
+
+// TODO: TestSetDotGlobalConfig - add cmd config tests
```
Comment (ryanchristo) [inline cmd/gossamer/config_test.go]: #685

File: cmd/gossamer/config_test.go [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+
+// TODO: TestSetDotAccountConfig - add cmd config tests
```
Comment (ryanchristo) [inline cmd/gossamer/config_test.go]: #685

File: cmd/gossamer/config_test.go [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+
+// TODO: TestSetDotCoreConfig - add cmd config tests
```
Comment (ryanchristo) [inline cmd/gossamer/config_test.go]: #685

File: cmd/gossamer/config_test.go [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+
+// TODO: TestSetDotNetworkConfig - add cmd config tests
```
Comment (ryanchristo) [inline cmd/gossamer/config_test.go]: #685

File: cmd/gossamer/config_test.go [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+
+// TODO: TestSetDotRPCConfig - add cmd config tests
```
Comment (ryanchristo) [inline cmd/gossamer/config_test.go]: #685

File: cmd/gossamer/export_test.go [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+
+package main
+
+// TODO: add cmd export tests
```
Comment (ryanchristo) [inline cmd/gossamer/export_test.go]: #685

File: cmd/gossamer/flags_test.go [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+
+package main
+
+// TODO: add cmd flag tests
```
Comment (ryanchristo) [inline cmd/gossamer/flags_test.go]: #685

File: cmd/gossamer/main_test.go [hunk 13]
```diff
@@ -183,17 +179,24 @@ func TestMain(m *testing.M) {
...
 		"failed to read command argument: \"potato\"",
 	}
+
 	for _, m := range expectedMessages {
 		require.Contains(t, gossamer.StderrText(), m)
 	}
-
 }
+
+// TODO: TestGossmaerCommand test "gossamer" does not error
```
Comment (ryanchristo) [inline cmd/gossamer/main_test.go]: #685

File: cmd/gossamer/main_test.go [hunk 14]
```diff
@@ -183,17 +179,24 @@ func TestMain(m *testing.M) {
...
+
+// TODO: TestExportCommand test "gossamer export" does not error
```
Comment (ryanchristo) [inline cmd/gossamer/main_test.go]: #685

File: cmd/gossamer/main_test.go [hunk 15]
```diff
@@ -183,17 +179,24 @@ func TestMain(m *testing.M) {
...
+
+// TODO: TestInitCommand test "gossamer init" does not error
```
Comment (ryanchristo) [inline cmd/gossamer/main_test.go]: #685

File: cmd/gossamer/main_test.go [hunk 16]
```diff
@@ -183,17 +179,24 @@ func TestMain(m *testing.M) {
...
+
+// TODO: TestAccountCommand test "gossamer account" does not error
```
Comment (ryanchristo) [inline cmd/gossamer/main_test.go]: #685

File: cmd/gossamer/main.go [hunk 17]
```diff
@@ -19,36 +19,44 @@ package main
...
-		Usage:       "Show configuration values",
-		ArgsUsage:   "",
-		Flags:       AllFlags(),
-		Category:    "CONFIGURATION DEBUGGING",
-		Description: `The dumpconfig command shows configuration values.`,
+	// exportCommand defines the "export" subcommand (ie, `gossamer export`)
+	exportCommand = cli.Command{
+		Action:    FixFlagOrder(exportAction),
+		Name:      "export",
+		Usage:     "Export configuration values to TOML configuration file",
```
Comment (ryanchristo) [inline cmd/gossamer/main.go]: #686

File: dot/utils_test.go [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+// TestNewConfigAndFile tests the NewTestConfigWithFile method
+func TestNewConfigAndFile(t *testing.T) {
+	testCfg, testCfgFile := NewTestConfigWithFile(t)
+
+	defer utils.RemoveTestDir(t)
+
+	require.NotNil(t, testCfg)     // TODO: improve dot utils tests
+	require.NotNil(t, testCfgFile) // TODO: improve dot utils tests
```
Comment (ryanchristo) [inline dot/utils_test.go]: #687

File: dot/utils_test.go [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+	"github.com/stretchr/testify/require"
+)
+
+// TestNewConfig tests the NewTestConfig method
+func TestNewConfig(t *testing.T) {
+	cfg := NewTestConfig(t)
+
+	defer utils.RemoveTestDir(t)
+
+	require.NotNil(t, cfg) // TODO: improve dot utils tests
```
Comment (ryanchristo) [inline dot/utils_test.go]: #687

File: dot/services_test.go [hunk 20]
```diff
@@ -0,0 +1,149 @@
...
+
+	ks := keystore.NewKeystore()
+
+	coreSrv, err := createCoreService(cfg, ks, stateSrv, networkMsgSend, networkMsgRec)
+	require.Nil(t, err)
+
+	rpcSrv := createRPCService(cfg, stateSrv, networkSrv, coreSrv)
+	require.Nil(t, err)
+
+	// TODO: improve dot service tests
```
Comment (ryanchristo) [inline dot/services_test.go]: #687

File: dot/services_test.go [hunk 21]
```diff
@@ -0,0 +1,149 @@
...
+func TestCreateRPCService(t *testing.T) {
+	cfg := NewTestConfig(t)
+	require.NotNil(t, cfg)
+
+	genFile := NewTestGenesisFile(t, cfg)
+	require.NotNil(t, genFile)
+
+	defer utils.RemoveTestDir(t)
+
+	cfg.Core.Authority = false // TODO: improve dot service tests
```
Comment (ryanchristo) [inline dot/services_test.go]: #687

File: dot/services_test.go [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+	_, networkMsgSend, networkMsgRec := createNetworkService(cfg, stateSrv)
+	require.Nil(t, err)
+
+	ks := keystore.NewKeystore()
+	require.NotNil(t, ks)
+
+	coreSrv, err := createCoreService(cfg, ks, stateSrv, networkMsgSend, networkMsgRec)
+	require.Nil(t, err)
+
+	// TODO: improve dot service tests
```
Comment (ryanchristo) [inline dot/services_test.go]: #687

File: dot/services_test.go [hunk 23]
```diff
@@ -0,0 +1,149 @@
...
+func TestCreateCoreService(t *testing.T) {
+	cfg := NewTestConfig(t)
+	require.NotNil(t, cfg)
+
+	genFile := NewTestGenesisFile(t, cfg)
+	require.NotNil(t, genFile)
+
+	defer utils.RemoveTestDir(t)
+
+	cfg.Core.Authority = false // TODO: improve dot service tests
```
Comment (ryanchristo) [inline dot/services_test.go]: #687

File: dot/services_test.go [hunk 24]
```diff
@@ -0,0 +1,149 @@
...
+	err := InitNode(cfg)
+	require.Nil(t, err)
+
+	stateSrv, err := createStateService(cfg)
+	require.Nil(t, err)
+
+	networkSrv, networkMsgSend, networkMsgRec := createNetworkService(cfg, stateSrv)
+	require.Nil(t, err)
+
+	// TODO: improve dot service tests
```
Comment (ryanchristo) [inline dot/services_test.go]: #687

File: dot/services_test.go [hunk 25]
```diff
@@ -0,0 +1,149 @@
...
+
+	cfg.Global.Genesis = genFile.Name()
+
+	err := InitNode(cfg)
+	require.Nil(t, err)
+
+	stateSrv, err := createStateService(cfg)
+	require.Nil(t, err)
+
+	// TODO: improve dot service tests
```
Comment (ryanchristo) [inline dot/services_test.go]: #687

File: dot/node_test.go [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+	go node.Start()
+	<-node.IsStarted
+
+	node.Stop()
+	<-node.stop
+}
+
+// TestStopNode
+
+// TODO: improve dot node tests
```
Comment (ryanchristo) [inline dot/node_test.go]: #687

File: dot/node_test.go [hunk 27]
```diff
@@ -0,0 +1,102 @@
...
+func TestStartNode(t *testing.T) {
+	cfg := NewTestConfig(t)
+	require.NotNil(t, cfg)
+
+	genFile := NewTestGenesisFile(t, cfg)
+	require.NotNil(t, genFile)
+
+	defer utils.RemoveTestDir(t)
+
+	cfg.Core.Authority = false // TODO: improve dot node tests
```
Comment (ryanchristo) [inline dot/node_test.go]: #687

File: dot/node_test.go [hunk 28]
```diff
@@ -0,0 +1,102 @@
...
+func TestNewNode(t *testing.T) {
+	cfg := NewTestConfig(t)
+	require.NotNil(t, cfg)
+
+	genFile := NewTestGenesisFile(t, cfg)
+	require.NotNil(t, genFile)
+
+	defer utils.RemoveTestDir(t)
+
+	cfg.Core.Authority = false // TODO: improve dot node tests
```
Comment (ryanchristo) [inline dot/node_test.go]: #687

File: dot/node_test.go [hunk 29]
```diff
@@ -0,0 +1,102 @@
...
+
+	cfg.Global.Genesis = genFile.Name()
+
+	err := InitNode(cfg)
+	require.Nil(t, err)
+}
+
+// TestNodeInitialized
+
+// TODO: improve dot node tests
```
Comment (ryanchristo) [inline dot/node_test.go]: #687

File: dot/config_test.go [hunk 30]
```diff
@@ -17,56 +17,132 @@
...
+	cfg.Global.Genesis = genFile.Name()
 
-	err := LoadConfig(fp, cfg)
+	err := InitNode(cfg)
 	require.Nil(t, err)
+
+	err = LoadConfig(cfg, cfgFile.Name())
+	require.Nil(t, err)
+
+	// TODO: improve dot config tests
```
Comment (ryanchristo) [inline dot/config_test.go]: #687

File: dot/config_test.go [hunk 31]
```diff
@@ -17,56 +17,132 @@
...
+	defer utils.RemoveTestDir(t)
+
+	cfg.Global.Genesis = genFile.Name()
+
+	err := InitNode(cfg)
+	require.Nil(t, err)
+
+	file := ExportConfig(cfg, cfgFile.Name())
+
+	// TODO: improve dot config tests
```
Comment (ryanchristo) [inline dot/config_test.go]: #687

File: dot/config_test.go [hunk 32]
```diff
@@ -17,56 +17,132 @@
...
+
+	defer utils.RemoveTestDir(t)
+
+	err := InitNode(cfg)
+	require.Nil(t, err)
+
+	err = LoadConfig(cfg, "../node/gssmr/config.toml")
+	require.Nil(t, err)
+
+	// TODO: improve dot config tests
```
Comment (ryanchristo) [inline dot/config_test.go]: #687

File: dot/config_test.go [hunk 33]
```diff
@@ -17,56 +17,132 @@
...
+	cfg.Global.Genesis = "../node/gssmr/genesis.json"
+
+	defer utils.RemoveTestDir(t)
+
+	err := InitNode(cfg)
+	require.Nil(t, err)
+
+	file := ExportConfig(cfg, "../node/gssmr/config.toml")
+
+	// TODO: improve dot config tests
```
Comment (ryanchristo) [inline dot/config_test.go]: #687

File: dot/config_test.go [hunk 34]
```diff
@@ -17,56 +17,132 @@
...
+	cfg.Global.Genesis = "../node/ksmcc/genesis.json"
+
+	defer utils.RemoveTestDir(t)
+
+	err := InitNode(cfg)
+	require.Nil(t, err)
+
+	err = LoadConfig(cfg, "../node/ksmcc/config.toml")
+
+	// TODO: improve dot config tests
```
Comment (ryanchristo) [inline dot/config_test.go]: #687

File: dot/config_test.go [hunk 35]
```diff
@@ -17,56 +17,132 @@
...
+
+	err := InitNode(cfg)
+	require.Nil(t, err)
 
-	fp := "../tests/exports/config.toml"
-	cfg := NewTestConfig(dir)
+	file := ExportConfig(cfg, "../node/ksmcc/config.toml")
 
-	file := ExportConfig(fp, cfg)
+	// TODO: improve dot config tests
```
Comment (ryanchristo) [inline dot/config_test.go]: #687

File: dot/node.go [hunk 36]
```diff
@@ -0,0 +1,239 @@
...
+	if os.IsNotExist(err) {
+		log.Warn(
+			"[dot] Node has not been initialized",
+			"datadir", cfg.Global.DataDir,
+			"error", "failed to locate MANIFEST file in data directory",
+		)
+		return false
+	}
+
+	// TODO: investigate cheap way to confirm valid genesis data has been loaded
```
Comment (ryanchristo) [inline dot/node.go]: #688

File: lib/utils/utils_test.go [hunk 37]
```diff
@@ -0,0 +1,19 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+
+package utils
+
+// TODO: add utils tests
```
Comment (ryanchristo) [inline lib/utils/utils_test.go]: #689

File: lib/utils/test_utils_test.go [hunk 38]
```diff
@@ -0,0 +1,19 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+
+package utils
+
+// TODO: add utils tests
```
Comment (ryanchristo) [inline lib/utils/test_utils_test.go]: #689

File: lib/utils/test_utils.go [hunk 39]
```diff
@@ -0,0 +1,53 @@
...
+//
+// The gossamer library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+
+package utils
```
Comment (ryanchristo) [inline lib/utils/test_utils.go]: #689

File: dot/node.go [hunk 40]
```diff
@@ -0,0 +1,239 @@
...
+	// State Service
+
+	// create state service and append state service to node services
+	stateSrv, err := createStateService(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create state service: %s", err)
+	}
+	nodeSrvcs = append(nodeSrvcs, stateSrv)
+
+	// TODO: Configure node based on Roles #601
```
Comment (ryanchristo) [inline dot/node.go]: #601

File: dot/genesis_test.go [hunk 41]
```diff
@@ -101,32 +133,35 @@ func TestStoreGenesisInfo(t *testing.T) {
...
+	require.NotNil(t, cfg)
+
+	genPath := newTestGenesisAndRuntime(t)
+	require.NotNil(t, genPath)
+
+	defer utils.RemoveTestDir(t)
 
-	genesisPath := createTempGenesisFile(t)
-	defer os.Remove(genesisPath)
+	cfg.Core.Authority = false // TODO: improve dot genesis tests
```
Comment (ryanchristo) [inline dot/genesis_test.go]: #687

**Comment by ryanchristo** [comment]
cc @thomasmodeneis opened #689 to address the utils tests and fixed the linting error :+1:

File: cmd/gossamer/export.go [hunk 42]
```diff
@@ -0,0 +1,75 @@
...
+	"os"
+	"path/filepath"
+
+	log "github.com/ChainSafe/log15"
+	"github.com/naoina/toml"
+	"github.com/urfave/cli"
+)
+
+// exportAction is the action for the "export" subcommand
+func exportAction(ctx *cli.Context) error {
```
Comment (noot) [inline cmd/gossamer/export.go]: what is this? was this dumpConfig before?

**Comment by noot** [comment]
when running a node i see an error:
```
elizabeth@elizabeth:~/go/src/github.com/ChainSafe/gossamer$ ./bin/gossamer --key alice 
INFO[03-11|13:51:41] [cmd] Loading node implementation...     name=gssmr
INFO[03-11|13:51:41] [dot] Setting up node services...        datadir=/home/elizabeth/.gossamer/gssmr
INFO[03-11|13:51:41] [dot] Creating state service... 
INFO[03-11|13:51:41] [dot] Creating network service...        port=7001 bootnodes=[] protocol=/gossamer/gssmr/0 roles=1 nobootstrap=false nomdns=false
WARN[03-11|13:51:41] [network] Bootstrap is enabled but no bootstrap nodes are defined 
INFO[03-11|13:51:41] [dot] Creating core service...           authority=true
INFO[03-11|13:51:42] [babe] config                            SlotDuration (ms)=1000 EpochLength (slots)=6
INFO[03-11|13:51:42] [dot] Creating rpc service...            host=localhost port=8545 mods="[system author]"
INFO[03-11|13:51:42] [cmd] Starting node...                   name=gssmr
INFO[03-11|13:51:42] [dot] Starting node services... 
INFO[03-11|13:51:42] Starting services: [*state.Service *network.Service *core.Service *rpc.HTTPServer] 
EROR[03-11|13:51:42] Error starting service                   srvc=*state.Service err="Cannot acquire directory lock on \"/home/elizabeth/.gossamer/gssmr\".  Another process is using this Badger database.: resource temporarily unavailable"
```

it looks concerning, i've tried deleting datadir/re-initializing and it's still there. could you fix this?

File: cmd/gossamer/export.go (hunk 42 above)
Comment (ryanchristo) [reply cmd/gossamer/export.go]: correct, it was `dumpConfig` before

**Comment by ryanchristo** [comment]
cc @noot that was introduced in a recent merged pull request and is appearing on `development`

**Comment by noot** [comment]
@ryanchristo okay that's bad... do you want to fix it here or should we just fix it on development first?

also, network tests hang forever when i try to run them locally

**Comment by ryanchristo** [comment]
I will look into the network error but the error on `development` is likely unrelated and it looks more like something happening in `core` based on your recent merge pull request

**Comment by ryanchristo** [comment]
I was not encountering either error before the last merge. You can test if you would like.

**Comment by noot** [comment]
okay, I'll investigate the DB issue. network tests still hang for me locally, is that happening for anyone else?

also, looks like the authority flag isn't working:
```
elizabeth@elizabeth:~/go/src/github.com/ChainSafe/gossamer$ ./bin/gossamer --key alice --authority=false
INFO[03-11|15:10:39] [cmd] Loading node implementation...     name=gssmr
INFO[03-11|15:10:39] [dot] Setting up node services...        datadir=/home/elizabeth/.gossamer/gssmr
INFO[03-11|15:10:39] [dot] Creating state service... 
INFO[03-11|15:10:39] [dot] Creating network service...        port=7001 bootnodes=[] protocol=/gossamer/gssmr/0 roles=1 nobootstrap=false nomdns=false
WARN[03-11|15:10:39] [network] Bootstrap is enabled but no bootstrap nodes are defined 
INFO[03-11|15:10:39] [dot] Creating core service...           authority=true
```

this should override the config option and set authority to false

**Comment by ryanchristo** [comment]
fixing the authority flag and looking into the network stall :+1:

**Comment by ryanchristo** [comment]
The hanging network tests were related to a `TODO` I had to remove a line after merging your pull request fixing message channels that I forgot about. Remove and fixed.

The flags have also been updated and tested. :+1:

**Comment by ryanchristo** [comment]
nevermind... looks like there are still some network errors. Looking into it. Flags fixed though!

**Comment by ryanchristo** [comment]
cc @noot the network tests are hanging on `development` for me now too. I'm not sure it's related to my pull request... can you investigate on `development` if it's happening for you there as well and if you are ok with the flag fixes, an approval.

**Comment by noot** [comment]
@ryanchristo yeah you're right, they are hanging for me too now. wondering why it didn't happen before :/ I will investigate

**Comment by ryanchristo** [comment]
oof, looking into it

**Comment by noot** [comment]
@ryanchristo I've added fixes for the 2 issues: https://github.com/ChainSafe/gossamer/pull/694

**Comment by ryanchristo** [comment]
cc @noot looks like tests are passing now :+1:

File: dot/node.go [hunk 43]
```diff
@@ -0,0 +1,241 @@
...
+// Node is a container for all the components of a node.
+type Node struct {
+	Name      string
+	Services  *services.ServiceRegistry // Registry of all core services
+	IsStarted chan struct{}             // Signals node startup complete
+	stop      chan struct{}             // Used to signal node shutdown
+}
+
+// InitNode initialize a new dot node from the provided dot node configuration
+// and JSON formatted gensis file
```
Comment (noot) [inline dot/node.go]: genesis

File: dot/node.go [hunk 44]
```diff
@@ -0,0 +1,241 @@
...
+		"[dot] Node initialized",
+		"datadir", dataDir,
+		"genesis", genPath,
+	)
+
+	return nil
+}
+
+// NodeInitialized returns true if, within the configured data directory for the
+// node, the state database has been created and the gensis data has been loaded
```
Comment (noot) [inline dot/node.go]: genesis

File: node/gssmr/config.toml [hunk 45]
```diff
@@ -1,16 +1,27 @@
 [global]
-data-dir = "~/.gossamer/gssmr"
-roles = 1
+name = "gssmr"
+id = "gssmr"
+config = "./node/gssmr/config.toml"
+genesis = "./node/gssmr/genesis.json"
```
Comment (noot) [inline node/gssmr/config.toml]: this isn't a global option, it's specific to the init subcommand

File: cmd/gossamer/config.go [hunk 46]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+// setDotGlobalConfig sets dot.GlobalConfig using flag values from the cli context
+func setDotGlobalConfig(ctx *cli.Context, cfg *dot.GlobalConfig) {
+	// check --config flag and update node confifuration
+	if config := ctx.GlobalString(ConfigFlag.Name); config != "" {
+		cfg.Config = config
+	}
+
+	// check --genesis flag and update node confifuration
```
Comment (noot) [inline cmd/gossamer/config.go]: configuration

File: cmd/gossamer/config.go [hunk 47]
```diff
@@ -0,0 +1,232 @@
...
+	if genesis := ctx.GlobalString(GenesisFlag.Name); genesis != "" {
```
Comment (noot) [inline cmd/gossamer/config.go]: not really a global flag, we shouldn't be able to use this if the node is initialized, or unless we are doing `init`

**Comment by ryanchristo** [comment]
Ok, typos fixed.

Hmm... I still see the way it is here as an advantage because if someone wants to start a node using a custom genesis file, they would run `gossamer --genesis custom/genesis.json` and the node would initialize and start using the custom genesis configuration file.

If they start the node again passing the genesis file `gossamer --genesis custom/genesis.json`, the node is already initialized so the genesis flag would have no affect (the flag is just a configuration setting, not what determines if the node will be initialized), and if they started the node without the genesis file, the node would already be initialized so it would again have no effect.

I just added a warning if they run the initialize command on an already initialized node `gossamer --genesis custom/genesis.json` letting them know that they are reinitializing the node and the node will be reinitialized with the custom genesis file. If they ran `gossamer init`, they would receive the warning but the node would be reinitialized using the default gossamer genesis.

I can open an issue but not sure I understand what you would prefer.

**Pull request merged**